iwyu.cc

```diff
@@ -2585,11 +2603,17 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
   const UsingDecl* GetUsingDeclarationOf(const NamedDecl* decl,
                                          const DeclContext* using_context) {
-    // We look through all the using-decls of the given decl.  We
-    // limit them to ones that are visible from the decl-context we're
-    // currently in (that is, what namespaces we're in).  Of those, we
-    // pick the one that's in the same file as decl, if possible,
-    // otherwise we pick one arbitrarily.
+    // First, if we have a UsingShadowDecl, then we don't need to do anything
+    // because we can just directly return the using decl from that.
+    if (const UsingShadowDecl* shadow = DynCastFrom(decl))
+      return shadow->getUsingDecl();
+
+    // But, if we don't have a UsingShadowDecl, then we need to look through
+    // all the using-decls of the given decl.  We limit them to ones that are
+    // visible from the decl-context we're currently in (that is, what
+    // namespaces we're in), via the check through 'Encloses'. Of those, we
```

Comment:
I wonder if this should also include `Equals` as I mentioned on the other PR? See, `Encloses` only checks if the `using_context` is an "ancestor" of `using_decl_context`, not if they're the same. Then again, maybe the shadow decl checks cover this case?

I've also been thinking about this more generally, I wonder if what we really want to do is a proper lookup? Not sure how that works, though, but it might remove some boilerplate here. For future reference :-)